### PR TITLE
byte_extract lowering of string constants [blocks: #2068]

### PR DESCRIPTION
**Pull request by tautschnig**
These need to be handled like arrays of characters.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/lowering/byte_operators.cpp [hunk 1]
```diff
@@ -118,6 +119,26 @@ static exprt unpack_rec(
         array.copy_to_operands(op);
     }
   }
+  else if(src.id() == ID_string_constant)
```
Comment (smowton) [inline src/solvers/lowering/byte_operators.cpp:279]: I might be wrong, but I think these are not used by Java, so the commit message may need amending

**Review by smowton — COMMENTED** [review]
AFAIK this is all C, not Java stuff, as particularly shown by the unit test which exhibits strings with char size = 8 (Java chars are 16-bit)

File: src/solvers/lowering/byte_operators.cpp (hunk 1 above)
Comment (tautschnig) [reply src/solvers/lowering/byte_operators.cpp:279]: Amended to not have that unnecessary restriction, but I do see results of `git grep string_constantt` in `jbmc/` as well...

Comment (smowton) [reply src/solvers/lowering/byte_operators.cpp:279]: They are used for `input` and `output` steps only; the Java bytecode parser makes them to describe string literals, but those are later transformed into references to constant `java.lang.String` instances, so there are no free-floating constants in the same way as there are in C.

Comment (tautschnig) [reply src/solvers/lowering/byte_operators.cpp:279]: Ah, ok! I guess it must have been some C code where I had seen them in the wild.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9a67ec4d550f8f3ae6901845c485e058e4e64028).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101065754

**Pull request merged**